### ISSUE TITLE
Implement UserMessages parsing

### ISSUE THREAD

**Issue by 0BuRner** [EHVAG/demoinfo#45]
It could be great to be able to retrieve usermessages.
There are some info such as players ranking data that are really useful.

I tried to do it some time ago but wasn't able don't remember why. If you could take a closer look at it would be awesome.

**Comment by geniusap** [comment]
There is already some kind of support for this, you just need to enable SLOW_PROTOBUF, and add a few lines to UserMessageHandler... it is quite easy to get a CCSUsrMsg_ServerRankUpdate object with the old/new ranks... 

What I'm trying to do and cannot manage yet is to get the chat log. Any hints / ideas regarding that?

**Comment by 0BuRner** [comment]
Yeah I tought it would be easy too but I didn't manage to get it works... :s Do you have some code to share ?
You should open a new issue for chat logs ;)

**Comment by moritzuehling** [comment]
I'm looking into this, I won't make any promises. I'm currently pretty busy and can't really work on this parser.

**Comment by geniusap** [comment]
@0BuRner, within UserMessageHandler you can do something like this...

```c#
Type toParse = null;
if (Enum.IsDefined(typeof(ECstrike15UserMessages), userMessage.msg_type)) {
  ECstrike15UserMessages msg = (ECstrike15UserMessages) userMessage.msg_type;
  toParse = Assembly.GetExecutingAssembly().GetType("DemoInfo.Messages.CCSUsrMsg_" + msg.ToString().Substring(6));
}
ProtoBuf.IExtensible result;
using (var memstream = new MemoryStream(userMessage.msg_data))
  result = memstream.ReadProtobufMessage(toParse);
```

Then if/when the result is assignable to CCSUsrMsg_ServerRankUpdate, you'll get the data you need (either add your own events and handlers, or just print it out there).

Regarding the chat, I managed to get access to the data, but for MM demos it is unfortunately encrypted (probably via tv_encryptdata_key or tv_encryptdata_key_pub).

**Comment by 0BuRner** [comment]
Thanks geniusap!
Found this : http://pastebin.com/4CgdZVkF

**Comment by gocke** [comment]
Yeah thats from me, i didn't share it on purpose. It's not finished, not only is it buggy as hell, often the values are also wrong.

**Comment by 0BuRner** [comment]
Seems this guy was able to retrieve ranking info from demo file : https://github.com/bugdone/headshotbox

**Comment by moritzuehling** [comment]
Fixed with #65
